src/NHibernate/Loader/JoinWalker.cs

```diff
@@ -980,42 +980,96 @@ protected SqlStringBuilder WhereString(string alias, string[] columnNames, int b
 			{
 				// if not a composite key, use "foo in (?, ?, ?)" for batching
 				// if no batch, and not a composite key, use "foo = ?"
-				string tableAlias = GenerateAliasForColumn(alias, columnNames[0]);
-				InFragment inf = new InFragment().SetColumn(tableAlias, columnNames[0]);
+				var columnName = columnNames[0];
+
+				var tableAlias = GenerateAliasForColumn(alias, columnName);
+				var qualifiedName = !string.IsNullOrEmpty(tableAlias)
+					? tableAlias + StringHelper.Dot + columnName
+					: columnName;
+				
+				var whereString = new SqlStringBuilder(batchSize * 5);
+				whereString.Add(qualifiedName);
+				if (batchSize == 1)
+				{
+					whereString.Add("=").Add(Parameter.Placeholder);
+				}
+				else
+				{
+					bool added = false;
 
-				for (int i = 0; i < batchSize; i++)
-					inf.AddValue(Parameter.Placeholder);
+					whereString.Add(" in (");
+					for (var i = 0; i < batchSize; i++)
+					{
+						var value = Parameter.Placeholder;
+						if (added)
+						{
+							whereString.Add(StringHelper.CommaSpace);
+						}
 
-				return new SqlStringBuilder(inf.ToFragmentString());
```

Comment:
Maybe we should instead extend `InFragment`  with provided `SqStringBuilder` in ctor or add property to return underlying builder? Maybe the same can be done with `DisjunctionFragment`.

Also `InFragment` contains some logic regarding `null`/`not null` values handling. It's now missed - though not sure if it's applicable  here.

Reply:
> though not sure if it's applicable here.

Not applicable. The values are always parameter placeholder, which is never null.

Reply:
> Maybe we should instead extend InFragment with provided SqStringBuilder in ctor or add property to return underlying builder? Maybe the same can be done with DisjunctionFragment.

Current logic creates a lot of unnecessary objects and arrays, not just SqlStringBuilder.